BDMS-227-254: add schema validations to aquifer and geology schemas

ksmuczynski opened the pull request:
<!--
BDMS-227-254: add schema validations to aquifer and geology schemas
-->
### **Why**
_This PR addresses the following problem / context:_
- Data Integrity: Currently, there are no checks to ensure that depth intervals are logical (i.e., bottom_depth > top_depth) or that spatial boundaries are valid WKT strings.
- Missing Functionality: There was no Pydantic schema available to handle the creation of ThingGeologicFormationAssociation records.
- Loose Typing: Input strings for fields like primary_aquifer_type or lithology were not strictly validated against their respective controlled vocabularies, leading to potential data inconsistencies (e.g., casing mismatches).

### **How**
_Implementation summary - the following was changed / added / removed:_

**New Validators Module**: Created schemas/validators.py containing reusable mixins:
    - DepthIntervalMixin: Enforces that bottom_depth > top_depth and depths are non-negative.
    - GeometryMixin: Validates boundary fields
    - validate_enum_input: A helper function to validate inputs against Enums case-insensitively.

**Geologic Formations:**
- Added CreateThingGeologicFormationAssociation schema to schemas_geologic_formation.py using the DepthIntervalMixin.
- Updated CreateGeologicFormation to enforce uppercasing on formation_code and Enum validation on lithology.

**Aquifer Systems:**
- Updated CreateAquiferSystem in schemas_aquifer_system.py to enforce AquiferType and GeographicScale Enum validation.
- Added GeometryMixin to CreateAquiferSystem for WKT boundary validation.

### **Notes**
_Any special considerations, workarounds, or follow-up work to note?_
- **Design decisions:** Shared validation logic was placed in a dedicated schemas/validators.py file rather than __init__.py to maintain clean separation of concerns and prevent circular import issues.

ksmuczynski committed:
feat(schemas): add enum, depth and geometry validations and association create model

- Add `schemas/validators.py` with `DepthIntervalMixin` for logical depth checks, `GeometryMixin` for WKT validation, and `validate_enum_input` for validation.
- Add `CreateThingGeologicFormationAssociation` to `schemas_geologic_formation.py` to support creation of formation picks/stratigraphy.
- Update `CreateGeologicFormation` to enforce uppercase formation codes and validate WKT boundaries.
- Update `CreateAquiferSystem` to validate WKT boundaries.

jacob-a-brown commented on schemas/validators.py:
```diff
@@ -0,0 +1,76 @@
+
+    # 3. Fail if no match found
+    valid_options = [str(e.value) for e in enum_cls]
+    raise ValueError(f"Invalid value '{v}'. Must be one of: {', '.join(valid_options)}")
```
By setting a field to an enum Pydantic will automatically check if that field is one of the values in the enum's items. Because of this I don't think the function `validate_enum_input` is needed; let Pydantic check the input and raise errors if necessary. I also think that inputs should be strict for case-sensitivity as we are targeting specific strings.

ksmuczynski replied:
Agreed. Since we want to treat these vocabulary terms as strict identifiers, I'll remove the custom case-insensitive logic and rely on Pydantic's native strict validation.

jacob-a-brown commented on schemas/validators.py:
```diff
@@ -0,0 +1,76 @@
+    raise ValueError(f"Invalid value '{v}'. Must be one of: {', '.join(valid_options)}")
+
+
+class DepthIntervalMixin(BaseModel):
```
This would be helpful for well screens, too, since depths are also checked there. Should we change the names of the well screen fields to match here? and reuse the validator

ksmuczynski replied:
I'm not sure I'm following your suggestion. Are you saying the DepthIntervalMixin could be used to validate well screens and that the current well screen validation should inherit this mixin?

jacob-a-brown replied:
That's what I'm saying - since it checks depths it can be used in various places where depths are defined.

jacob-a-brown replied:
But we could add that to technical debt and keep it as-is for now

jacob-a-brown commented on schemas/validators.py:
```diff
@@ -0,0 +1,76 @@
+                    f"than top depth ({self.top_depth})"
+                )
+        if self.top_depth < 0:
+            raise ValueError("Top depth cannot be negative.")
```
numeric values should be checked and validated at the field itself using the [Field](https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.Field) object. For the individual field you can set (for example)

```
top_depth: float = Field(gt=0)
```

You can also replicate this with bottom depth for greater than or equal

```
bottom_depth: float = Field(ge=0)
```

jacob-a-brown commented on schemas/validators.py:
```diff
@@ -0,0 +1,76 @@
+
+    @model_validator(mode="after")
+    def check_depth_logical_order(self) -> "DepthIntervalMixin":
+        if self.bottom_depth is not None and self.top_depth is not None:
```
since both `top_depth` and `bottom_depth` are required this is check isn't necessary

jacob-a-brown commented on schemas/validators.py:
```diff
@@ -0,0 +1,76 @@
+
+        # Basic String Check
+        if not isinstance(v, str) or not v.strip():
+            raise ValueError("Boundary must be a valid WKT string.")
```
this just checks that it's a string, but not a valid WKT. To do that you may need to coerce it into a geometry object and catch any errors. See **services/validation/geospatial.py** for how this is done (and it can be replicated). 

Also should that validator be moved to schemas?

ksmuczynski replied:
What do you mean by move the validator to the schemas? I thought it was in schemas?

jacob-a-brown replied:
I meant move that validator function from **services/validation/geospatial.py** to the schemas directory somewhere.

jacob-a-brown commented on schemas/geologic_formation.py:
```diff
@@ -1,22 +1,51 @@
+    """
+
+    thing_id: int
+    geologic_formation_id: int | None = None
```
shouldn't this be required?

ksmuczynski replied:
Yes!!!

ksmuczynski committed:
refactor(schemas): centralize validations and enforce strict enums

- Delegate WKT boundary validation to `services/validation/geospatial.py` to enforce topological validity.
- Update `CreateAquiferSystem` and `CreateGeologicFormation` to enforce strict Enum typing for controlled vocabularies, removing loose string coercion.

ksmuczynski committed:
refactor(schemas): remove None check on validators.DepthIntervalMixin

`top_depth` and `bottom_depth` are required so the None check is redundant.

Remove the None check.

ksmuczynski committed:
refactor(schemas): enforce non-negative depths in DepthIntervalMixin

- Remove manual `if` validation logic for non-negative depths in `DepthIntervalMixin`.
- Implement `Field(ge=0)` on `top_depth` and `bottom_depth` to leverage Pydantic's native schema validation and cleaner OpenAPI generation.
- Ensure `CreateThingGeologicFormationAssociation` inherits these constraints by explicitly redefining fields with `ge=0`.

ksmuczynski committed:
Merge remote-tracking branch 'origin/bdms-227' into kas-227-254-add-schema-validations

# Conflicts:
#	schemas/aquifer_system.py
